### ISSUE TITLE
Translating keras model to darknet format

### ISSUE THREAD

**Issue by maustin8** [AlexeyAB/darknet#2587]
Hello @AlexeyAB 

I have a very simple binary CNN model for image classification that I have implemented using keras that I am able to get fairly okay prediction accuracy on. I am looking to implement this same model using darknet due to its simplicity to integrate in a C++ application. My problem is that when I try to translate my keras model to the darknet format, I seem to always get biased results where one class is always predicted over the other. I am fairly new to darknet and since I know I can get balanced results using keras I have a feeling that these errors are due to how I created the configuration file. 

The structure of the layers is fairly straight forward so I believe that I have translated that correctly but I may not have the same network parameters set. My keras model uses the Adam Optimizer (which I believe I have implemented in darknet), sparse categorical crossentropy for the loss function, and uses 5 epochs with 100 steps per epoch. Is it possible to define these in a darknet configuration file? I see that epoch is a part of the network struct that can be configured but it does not seem to be utilized when training.

I am not sure if this is the reason why I am getting such different results but I would like to make sure that my networks are configured as closely as possible so I know where else to look for bugs.

Thank you.

**Comment by AlexeyAB** [comment]
@maustin8 Hi,

* Do you want to convert Keras-weights to Darknet-weights? or do you want to make Darknet-cfg the simmilar to Keras-model?

* Do you want to train your model by using Darknet?

* Can you show your Keras network structure that you want to translate to the Darknet?

* And if you already made some Darknet cfg-file, can you show it?

**Comment by KDHAP** [comment]
I have similar question
I want to convert keras model(.h5) to darknet model(.weights)
I have simple CNN classifier model

My keras network structure is

model.add(layers.Conv2D(16, (20, 5), activation='relu', input_shape=(400, 100, 3)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(32, (16, 4), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(48, (8, 2), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Flatten())
model.add(layers.Dense(51, activation='softmax'))
_______________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
conv2d_82 (Conv2D)           (None, 381, 96, 16)       4816      
_________________________________________________________________
max_pooling2d_79 (MaxPooling (None, 190, 48, 16)       0         
_________________________________________________________________
conv2d_83 (Conv2D)           (None, 175, 45, 32)       32800     
_________________________________________________________________
max_pooling2d_80 (MaxPooling (None, 87, 22, 32)        0         
_________________________________________________________________
conv2d_84 (Conv2D)           (None, 80, 21, 48)        24624     
_________________________________________________________________
max_pooling2d_81 (MaxPooling (None, 40, 10, 48)        0         
_________________________________________________________________
flatten_24 (Flatten)         (None, 19200)             0         
_________________________________________________________________
dense_33 (Dense)             (None, 51)                979251    
=================================================================
Total params: 1,041,491
Trainable params: 1,041,491
Non-trainable params: 0

My model can be easily created as a cfg file.

plz help me..
thanks

**Comment by AlexeyAB** [comment]
@KDHAP 

You should save to the Darknet `.weights`-file:

1. save darknet-header - the first 5 values (int 32-bit) with such content [0, 2, 5, 0, 0]: 
    * https://github.com/ultralytics/yolov3/blob/d7a28bd9f74d922216e06de3dde5f981b3002bd4/models.py#L185-L186 
     *  https://github.com/ultralytics/yolov3/blob/d7a28bd9f74d922216e06de3dde5f981b3002bd4/models.py#L329-L330

2. Then save weights parameters of each convolutional layers ( `bias, weight , running_mean , running_variance` ): https://github.com/ultralytics/yolov3/blob/d7a28bd9f74d922216e06de3dde5f981b3002bd4/models.py#L339-L342

**Comment by rshigemura** [comment]
Hi @AlexeyAB !!

I have a little keras model splited in 2 files, weights (.h5) and architecture (.json) and need to translate it to darknet, because it is much faster for inference (files here: [https://we.tl/t-a4PmlBIIzJ]). I tried hard to follow your instructions above, but could not make it work.

Can you point out some other way to make it, so i could try? Better yet, if you find useful and easy enough, I can sponsor you on write a script for keras -> darknet model translation,  that we could use it further.

Could you help us, please?
Thanks in advance

Rafael.
